disorder-fsm/src/Disorder/FSM/Core.hs

```diff
@@ -0,0 +1,124 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Disorder.FSM.Core (
+  -- * Transition type and constructor
+    Transition(..)
+  , mapTransition
+  , Action(..)
+  , mapAction
+  -- * Lift functions for 'Gen'
+  , liftGen
+  -- * Transition evaluation
+  , runFSM
+  -- * Transition evaluation
+  , mapPropertyM
+  , mapContT'
+  , pickGen
+  ) where
+
+import           Control.Applicative (Applicative)
+import           Control.Exception.Base (AssertionFailed(..))
+import           Control.Monad.Trans.Class (MonadTrans(..))
+import           Control.Monad (Monad(..))
+import           Control.Monad.Catch (MonadCatch(..), MonadThrow(..), catchAll)
+import           Control.Monad.Cont (ContT(..))
+import           Control.Monad.IO.Class (MonadIO(..))
+import           Control.Monad.Reader (MonadReader(..))
+import           Control.Monad.State (MonadState(..), evalStateT, modify)
+
+import           Data.Function ((.), ($), const)
+import           Data.Functor (Functor(..))
+import           Data.Maybe (Maybe(..))
+import           Data.Monoid (Monoid(..), (<>))
+import           Data.String (String)
```

Comment:
[Eric asked the same question previously](https://github.com/ambiata/disorder.hs/pull/42#discussion_r36814258) And the answer is still the same: these `String`'s are used only in `show` and `fail` so we would still need to convert `Text` to `String` while and with `OverloadedStrings` it shouldn't matter for a user.